jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -5307,4 +5307,29 @@ public IntStream codePoints() {
 		return CharSequence.super.codePoints();
 	}
 /*[ENDIF]*/	
+
+/*[IF Java10]*/
+	/*
+	 * Internal API, assuming no modification to the byte array returned.
+	 * In case of enableSharingInSubstringWhenOffsetIsZero is set to true (by default), the actual bytes will be 
+	 * copied to a new array, otherwise original byte array value is returned.
+	 */
+	byte[] value() {
+		int	validLen;
+		if (enableCompression) {
+			if (LATIN1 == coder) {
+				validLen = count;
+			} else {
+				validLen = (count & ~uncompressedBit) * 2;
+			}
+		} else {
+			validLen = count * 2;
+		}
+		if (validLen < value.length) {
+			return java.util.Arrays.copyOf(value, validLen);
```

Comment:
if we are doing this copy we may want to consider updating value to the new array so that repeated calls to value do not have to copy the array repeatedly...

Reply:
So String.value can be modified using unsafe and this is done as part of the deduplication logic in called from String.equals. The same Unsafe method could be used to update the value field of the string with this shorter array if we wanted to. String objects have to always represent the same string, but since the two arrays are identical in the range code will be examining swapping them is safe (it is already done in equals). I am not certain we want to do this, but it may be necessary for performance depending on how often value is being called from outside the string class. Note, I'm not sure what the design motivation in Java 10 is for this API but it seems to be increasing the coupling between the JCL and the internal details of String which doesn't seem like a good design idea in general IMO.

Reply:
We're going to have problems with these types of methods moving forward as well as the use of `coder` as it seems it gets used quite often in the JCL after Java 9.

Reply:
Updated the PR to set `String.value` with newly created byte array. Pls review again.

Note about usage of this `Java 10` API `j.l.String.value()`, it is invoked by another `Java 10` method `j.l.StringCoding.getBytesUTF8NoRepl(string)` which is referred by `util.zip.ZipCoder`.